src/Parser/specialized.jl

```diff
@@ -0,0 +1,122 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start. Does not do full error checking.
+    fastpath, len = predict_string(ps)
+
+    # Now read the string itself
+    b = @compat Array{UInt8, 1}(len)
+
+    # Fast path occurs when the string has no escaped characters. This is quite
+    # often the case in real-world data, especially when keys are short strings.
+    # We can just copy the data from the buffer in this case.
+    if fastpath
+        s = ps.s
+        unsafe_copy!(b, 1, ps.utf8data, s + 1, len)
+        ps.s = s + len + 2
+    else
+        parse_string(ps, b)
+    end
+
+    _String(b)
+end
+
+function predict_string(ps::MemoryParserState)
```

Comment:
It's hard to make the comment short, since what the function does is a little strange. I added a docstring for this and the following function.